sky/skylet/providers/gcp/config.py

```diff
@@ -504,24 +504,45 @@ def _check_firewall_rules(vpc_name, config, compute):
         return False
     effective_rules = response["firewalls"]
 
-    def _get_refined_rule(rule):
-        KEY_TO_COMPARE = {"sourceRanges", "allowed", "direction"}
-        refined_rule = {}
-        for k in KEY_TO_COMPARE:
-            if k not in rule:
-                continue
-            if k == "allowed":
-                refined_rule[k] = sorted(rule[k], key=lambda x: x["IPProtocol"])
-            else:
-                refined_rule[k] = rule[k]
-        return refined_rule
-
-    required_rules = list(map(_get_refined_rule, required_rules))
-    effective_rules = list(map(_get_refined_rule, effective_rules))
-
-    for rule in required_rules:
-        if rule not in effective_rules:
+    def _merge_and_refine_rule(rules):
+        # Output: dict of (direction, source) -> {protocol -> ports}
+        tmp = {}
```

Comment:
How about rename `tmp` to `source2allowed_list`? Also, it would be great to have a type annotation for it: `Dict[Tuple[str, str], List[Dict[str, str]]`

Reply:
Yes, renamed!